### PR TITLE
Add wappalyzer dom detection and unit test about it

### PR DESCRIPTION
**Pull request by Darkiros**
Add wappalyzer dom detection and unit test about it

### CODE REVIEW

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 1]
```diff
@@ -129,6 +139,21 @@ def normalize_application_regex(self):
 
                     for i, pattern in enumerate(self.applications[application_name][dict_field][key]):
                         self.applications[application_name][dict_field][key][i] = self.normalize_regex(pattern)
+            
+            # Format each dom of applications
+            tmp_dom = []
+            for (key, patterns) in self.applications[application_name]["dom"].items():
+                #print(self.normalize_application_regex_dom({key:patterns}))
```
Comment (Slokilla) [inline wapitiCore/wappalyzer/wappalyzer.py:146]: Ça ressemble à du debug ça :p

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 2]
```diff
@@ -320,6 +349,59 @@ def detect_application_versions(self, application: dict) -> Set[str]:
 
         return versions
 
+    def detect_versions_normalize_dom(self, application: dict) -> Set[str]:
+        versions = set()
+        soup = self.html.soup
+        #print(application["dom"])
```
Comment (Slokilla) [inline wapitiCore/wappalyzer/wappalyzer.py:355]: Ça ressemble à du debug ça :p

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 3]
```diff
@@ -320,6 +349,59 @@ def detect_application_versions(self, application: dict) -> Set[str]:
...
+        for element in application["dom"]:
+            #print(element)
```
Comment (Slokilla) [inline wapitiCore/wappalyzer/wappalyzer.py:357]: Ça ressemble à du debug ça :p

**Review by Slokilla — CHANGES_REQUESTED** [review]

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 4]
```diff
@@ -307,6 +332,10 @@ def detect_application_versions(self, application: dict) -> Set[str]:
             "html": self._html_code,
             "scriptSrc": self._scripts,
         }
+
+        # Detect version of dom element
+        versions.update(self.detect_versions_normalize_dom(application))
```
Comment (bretfourbe) [inline wapitiCore/wappalyzer/wappalyzer.py:337]: Can you move this before defining `elements_to_check` or after the last `versions.update` call ?

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 5]
```diff
@@ -129,6 +139,21 @@ def normalize_application_regex(self):
...
+                tmp_dom.append(self.normalize_application_regex_dom({key:patterns}))
+            self.applications[application_name]["dom"] = tmp_dom
+
+    def normalize_application_regex_dom(self, dom : dict) -> dict:
+        for css_selector, value in dom.items():
+            # if contains nothing we just want to match with the css selector
+            if value == "":
+                return {css_selector : {"exists": ""}}
+            else:
+                return dom
```
[the remark this answers is not in the document]
Comment (Darkiros) [reply wapitiCore/wappalyzer/wappalyzer.py:156]: > You have a for loop that always returns during the first loop. Something is strange here

Is it better with something like this ?

```
css_selector = list(dom.keys())[0]
        value = dom[css_selector]
        # if contains nothing we just want to match with the css selector
        if value == "":
            return {css_selector : {"exists": ""}}
        else:
            return dom     
```

[the remark this answers is not in the document]
Comment (polyedre) [reply wapitiCore/wappalyzer/wappalyzer.py]: I do not know exactly what you want to do here, but it seems to me that you are trying to convert the data from Wappalyzer to a generic format used by Wapiti.

But by reading the function, if the first css_selector has no value, you return an empty match rule.

Please explain what this function does, and describe briefly its purpose in a comment inside the code.

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 6]
```diff
@@ -318,8 +349,66 @@ def detect_application_versions(self, application: dict) -> Set[str]:
...
+                    if item.get(attribute):
+                        versions.add("__detected__")
+            # Else we have to get value inside the attribute and it must match with the regex
+            else:
+                regex = ApplicationData.normalize_regex(value)
+                for item in match:
+                    if attribute == "text":
+                        result = item.getText()
+                    else:
+                        result = item.get(attribute)
```
Comment (polyedre) [inline wapitiCore/wappalyzer/wappalyzer.py:407]: Can you rename "item" and "result" to more explicit variable names ?

File: wapitiCore/wappalyzer/wappalyzer.py [hunk 7]
```diff
@@ -129,6 +140,25 @@ def normalize_application_regex(self):
...
+        Convert the result of wappalyzer to a generic format used by Wapiti.
+        The goal is to match with the css selector if we don't have a regex.
+        In that case we set an empty value else we put the regex.
+        """
+        css_selector = list(dom.keys())[0]
+        value = dom[css_selector]
+        if value == "":
+            return {css_selector : {"exists": ""}}
+        else:
+            return dom
```
Comment (polyedre) [inline wapitiCore/wappalyzer/wappalyzer.py:161]: Here, what if dom contains multiple regular expressions? The 'first' could be empty but not the others

**Review by polyedre — CHANGES_REQUESTED** [review]
This is better!

File: wapitiCore/wappalyzer/wappalyzer.py (hunk 7 above)
Comment (Darkiros) [reply wapitiCore/wappalyzer/wappalyzer.py:161]: There is only one. I do that to fit with the wapiti format which is list.

**Review by polyedre — APPROVED** [review]

**Comment by fwininger** [comment]
@Darkiros peux tu corriger les erreurs de coding style. Merci

**Comment by bretfourbe** [comment]
@fwininger @devl00p

**Pull request merged**